### DISCUSSION TITLE
unable to access dynamic attributes

### DISCUSSION THREAD

**Discussion by edenhanan** [seleniumbase/SeleniumBase#2309] [Q&A] [answered]
![image](https://github.com/seleniumbase/SeleniumBase/assets/63960881/2275fe89-5d97-4dee-9eca-8785d89ceb85)

```html
<button id="__xmlview0--objectPage-anchBar-__section1-__xmlview0--objectPage-2-anchor-internalSplitBtn-arrowButton" 
data-sap-ui="__xmlview0--objectPage-anchBar-__section1-__xmlview0--objectPage-2-anchor-internalSplitBtn-arrowButton" 
data-sap-ui-render="" 
data-ui5-accesskey="" 
aria-label="פתח תפריט" 
class="sapMBtnBase sapMBtn sapMSBArrow" 
tabindex="-1" 
aria-expanded="true" 
role="none" 
aria-controls="__menu12">
</button>
```
I have a button element that opens a menu, the only connection between them in the HTML is an attribute called  **aria-controls** that holds the ID of the menu(I cannot know the ID in advance as it is dependent on the number of menus in the screen) but that attribute only appears when the menu is already expanded. 

```python

self.click('//div[@aria-posinset="2"]//div[@class="sapMSBInner"]//button[@aria-label="פתח תפריט"]')
menu_id = self.get_attribute(
'//div[@aria-posinset="2"]//div[@class="sapMSBInner"]//button[@aria-label="פתח תפריט"]', 'aria-controls')
```

I tried the code above to get the value from "aria-controls" but got an error that the attribute does not exist 

this is the menu element for reference
```html
<div id="__menu12" 
data-sap-ui="__menu12" 
data-sap-ui-render="" 
tabindex="-1" 
hidefocus="true" 
class="sapUiMnu sapUiMnuCozySupport sapMMenu sapUiShd" 
data-sap-ui-popup="id-1700648347842-65" >
<ul role="menu" class="sapUiMnuLst sapUiMnuNoIco sapUiMnuNoSbMnu"></ul>
</div>

**Comment by mdmintz** [comment, marked as answer]
Have you tried simplifying the selector to `'button[aria-label="פתח תפריט"]'`? (That should work as a CSS Selector)

```python
attribute = self.get_attribute('button[aria-label="פתח תפריט"]', 'aria-controls')
```

Or if it's already visible, you can use this to click it:

```python
self.click('button[aria-label="פתח תפריט"]')
```

If it's hidden, you can use `js_click()` to click it:

```python
self.js_click('button[aria-label="פתח תפריט"]')
```

Whenever there's a hidden menu item under a dropdown menu, use `self.js_click(selector)` to click it directly.

**Reply by edenhanan** [reply]
hey thank you  for the fast replay
yes I tried but unfortunately, there are a couple of elements with this "aria-label"
![image](https://github.com/seleniumbase/SeleniumBase/assets/63960881/079484bf-aa05-451a-8077-196f6527311b)

**Reply by mdmintz** [reply]
It may help to use `:nth-of-type(n)` in the selector. For example:

```python
self.js_click('button[aria-label="פתח תפריט"]:nth-of-type(1)')

self.js_click('button[aria-label="פתח תפריט"]:nth-of-type(2)')

self.js_click('button[aria-label="פתח תפריט"]:nth-of-type(3)')
```

Without seeing the rest of the html, it is difficult to find the best selector.

Since `self.js_click(selector)` also works on hidden elements, you could also try clicking the final button that you want directly.

**Reply by edenhanan** [reply]
the problem is that the menu element and the attribute "aria-controls" are generated after I click this button element "//div[@aria-posinset="2"]//div[@class="sapMSBInner"]//button[@aria-label="פתח תפריט"]".
and when I try to get the value in the attribute "aria-controls", I get an error "attribute not found" or the value "None" when I know the attribute exists.

for example, I tried the selector you suggested with find_elements() and iterated through the list of elements to find the "aria-controls" attribute but got an empty list:
![image](https://github.com/seleniumbase/SeleniumBase/assets/63960881/139a5891-a55e-450b-9521-d264b67d3219)
![image](https://github.com/seleniumbase/SeleniumBase/assets/63960881/85f3aa12-c3fd-4f2c-9b17-0dbf14857ad7)

another example I tried using self.get_attribute() but got an error
```python

self.click('//div[@aria-posinset="2"]//div[@class="sapMSBInner"]//button[@aria-label="פתח תפריט"]')
menu_id = self.get_attribute(
'//div[@aria-posinset="2"]//div[@class="sapMSBInner"]//button[@aria-label="פתח תפריט"]', 'aria-controls')
```
```
Exception: Element {//div[@aria-posinset="2"]//div[@class="sapMSBInner"]//button[@aria-label="פתח תפריט"]} has no attribute {aria-controls}!
```

if it helps by any chance this is the element with the attribute I'm trying to get
![image](https://github.com/seleniumbase/SeleniumBase/assets/63960881/e77f2090-9029-408c-b60d-e1513e5e3c71)

**Reply by mdmintz** [reply]
Let the [SeleniumBase Recorder](https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/recorder_mode.md) generate the script for you. It might find a better selector to use.

Also, if there's text in the button, then you can also use SeleniumBase's special `TAG:contains("TEXT")` selector. Example:

```python
self.click('button:contains("Tel Aviv")')
```

**Reply by edenhanan** [reply]
thank you 
eventually, i found if I use find_visible_elements() instead of find_elements() I get only the open list
```python
self.click('//div[@aria-posinset="2"]//div[@class="sapMSBInner"]//button[@aria-label="פתח תפריט"]')
            menu_list_2 = self.find_visible_elements('//ul[@role="menu"]//li')
            for i in menu_list_2:
                if i.text == 'personal-info':
                    i.click()
                    break
```
